website/blog/2024-09-23-dangerzone.md

```diff
@@ -0,0 +1,193 @@
+
+By doing so, Dangerzone now has two containers with different responsibilities:
+* The **outer** Docker/Podman container acts as the **portability** layer for Dangerzone. Its main responsibility is to bundle the necessary config files, scripts, and programs to run gVisor. It's also responsible for bundling the container image that gVisor will spawn a container from.
+* The **inner** gVisor container acts as the **isolation** layer for Dangerzone. Its sole responsibility is to run the actual Dangerzone logic for rendering documents to pixels.
+
+<figure>
+<img src="/assets/images/2024-09-23-dangerzone-with-gvisor.svg" alt="Diagram showing the Dangerzone UI sending a document to a document renderer within an inner container, which is protected by gVisor’s Sentry. The Sentry intercepts system calls, allowing only limited system calls to pass to the Linux kernel with strict security settings. I/O system calls are handled by gVisor Gofer in an outer container, with less strict but controlled permissions">
+<figcaption>Outline of how gVisor integrates with Dangerzone. There are now two nested containers, and each one brings its own protections. Usage of LibreOffice is implied.</figcaption>
+</figure>
+
+Running gVisor inside a container came with its own set of challenges:
+* The Docker/Podman’s seccomp filter must allow the `ptrace` system call. We found that recent Docker Desktop versions and Podman version >= 4.0 have a seccomp filter that allows this system call. For older versions, we specified a custom seccomp filter that allowed it.
+* gVisor cannot run under SELinux in enforcing mode under default settings, so we labeled the container with `container_engine_t` (see GitHub issue [#880](https://github.com/freedomofpress/dangerzone/issues/880)).
+* The Docker/Podman container must run with the `SYS_CHROOT` capability. This is needed by gVisor to restrict its own access to the filesystem before it starts document processing. Other than that, the **outer** container drops all other capabilities and privileges.
+
+> 💡 You can find more details about this integration in the Dangerzone’s [gVisor design doc](https://github.com/freedomofpress/dangerzone/blob/main/docs/developer/gvisor.md).
+## Dangerzone protections
+We talked about Dangerzone’s original attack surface, and how we integrated gVisor to reduce it. In practice though, in what ways is Dangerzone better off than before? Well, if the Matryoshka containers are giving you a headache, or you just skimmed to this section (no shade), here’s how the new Dangerzone protections fare against the previous version, and the default protections of Linux containers:
```

Comment:
Nit, there are a few missing newlines here.

(and yes, I've diffed your markdown against ours...)